src/couch_replicator_connection.erl

```diff
@@ -0,0 +1,183 @@
+    ok = config:listen_for_changes(?MODULE, self()),
+    Interval = config:get_integer("replicator", "connection_close_interval", ?DEFAULT_CLOSE_INTERVAL),
+    {ok, Timer} = timer:send_after(Interval, close_idle_connections),
+    process_flag(trap_exit, true),
+    {ok, #state{close_interval=Interval, timer=Timer}}.
+
+
+acquire(URL) ->
+    case gen_server:call(?MODULE, {acquire, URL}) of
+        {ok, Worker} ->
+            link(Worker),
+            {ok, Worker};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+
+relinquish(Worker) ->
+    gen_server:cast(?MODULE, {relinquish, Worker}).
+
+create_worker_async(URL) ->
+    spawn_link(fun() ->
+        {ok, Pid} = ibrowse:spawn_link_worker_process(URL),
+        exit({ok, Pid})
+    end).
+
+handle_call({acquire, URL}, From, State) ->
+    {Pid, _Ref} = From,
+    monitor(process, Pid),
+    case ibrowse_lib:parse_url(URL) of
+        #url{host=Host, port=Port} ->
+            case ets:match_object(?MODULE, #connection{host=Host, port=Port, owner=undefined, _='_'}, 1) of
+                '$end_of_table' ->
+                    AsyncPid = create_worker_async(URL),
+                    true = ets:insert_new(
+                        ?MODULE,
+                        #connection{host=Host, port=Port, owner=Pid, worker=AsyncPid}
+                    ),
+                    Waiting = dict:store(AsyncPid, From, State#state.waiting),
+                    {noreply, State#state{waiting=Waiting}};
+                {[Worker], _Cont} ->
+                    ets:insert(?MODULE, Worker#connection{owner=Pid}),
+                    {reply, {ok, Worker#connection.worker}, State}
+            end
+    end.
+
+
+handle_cast({relinquish, WorkerPid}, State) ->
+    case ets:lookup(?MODULE, WorkerPid) of
+        [Worker] ->
+            ets:insert(?MODULE, Worker#connection{owner=undefined});
+        [] ->
+            ok
+    end,
+    {noreply, State}.
+
+
+% owner crashed
+handle_info({'DOWN', _Ref, process, Pid, _Reason}, State) ->
+    Workers = ets:match_object(?MODULE, #connection{owner=Pid, _='_'}),
+    lists:foreach(fun(Worker) ->
+        ets:insert(?MODULE, Worker#connection{owner=undefined})
+    end, Workers),
+    {noreply, State};
+
+% worker crashed
+handle_info({'EXIT', Pid, Reason}, State) ->
+    case dict:find(Pid, State#state.waiting) of
+        {ok, From} ->
+            Response = case Reason of
+                {ok, WorkerPid} ->
+                    [Worker] = ets:lookup(?MODULE, Pid),
+                    ets:insert(?MODULE, Worker#connection{worker=WorkerPid}),
+                    {ok, WorkerPid};
+                Other ->
+                    couch_log:error("Async replication connection creator died: ~p", [Other]),
+                    error
+            end,
+            gen_server:reply(From, Response),
+            {noreply, State#state{waiting=dict:erase(Pid, State#state.waiting)}};
+        error ->
```

Comment:
Making the separateness of these two clauses explicit sounds good to me, but I generally prefer to express that that via a comment when a bit of code is only used in one place, otherwise it's just spreading logically coupled code through different parts of a module. I don't have a strong opinion though. @rnewson @nickva Thoughts?